### ISSUE TITLE
Come up with exclusion rules for fuzzy searching

### ISSUE THREAD

**Issue by apendleton** [mapbox/fuzzy-phrase#29]
Broadly we know that at least in the near-term we don't want to fuzzy-match house numbers (because virtually all numbers will be indexed so the match cardinality will be way high), or non-alphabetic text (because we don't know what fuzzy search means right now for Chinese). There are likely some other pathological conditions that we'll want to avoid trying to fuzzy-match, like very long words, which might be computationally expensive. We need to make these rules concrete.

## Numbers
* **Proposal:** don't fuzzy-index or attempt to fuzzy-match any words that contain a digit or a `#`.
* **Alternatives/considerations:**
    * This means we can't fuzzy-correct `3nd` to `3rd` even though the typo is a letter-typo rather than a number-typo. We could consider only excluding words comprised *exclusively* of numbers or `#`, but I think this would be pretty high-noise because of address numbers like `221B`.
    * If we're symmetric at index and match time, it means we can't correct typos that introduce a number into a word that doesn't have one in it. Like, `Main` typo'ed as `Ma1n` wouldn't work if we exclude words containing digits at query time even though we'd have the necessary index contents to do it. We could have separate rules for query time (like, only attempt if word contains at most `edit_distance` digits). I'm not sure if this is worth the effort or not -- there's some compute cost to even calculating the digit count.

## Alphabets
* **Proposal:** for fuzzy searching, support Latin, Cyrillic, and Greek, with no diacritics, as a hard-coded regex in the library.
* **Alternatives/considerations:**
    * We could support more alphabets (or abjads like Arabic, or abugidas like Devanagari). I'm sort of inclined to start conservatively though? Same with diacritics.
    * Rather than hard-coding, we could make this configurable by the caller. This would add complexity, but one big concern with hard-coding is that it makes changing the included set a breaking change (post-change, the library will look in old indexes expecting to be able to find alphabets that weren't included in there at index time).
        * Or, an in-between option: we don't allow configuration at indexing time, but we do have the indexer write its current hardcoded configuration into a configuration file which is dynamically loaded for queries, so that old indexes' build-time configuration will be honored.
        * Yet another variant: we could have a regex that limits what can be indexed, but the thing we write to the config file captures only the characters that actually occur in real index entries (so, us-address captures that it only contains Latin letters). This might be beneficial for query time: don't try to spelling-correct Cyrillic words at in us-address, etc. Or maybe don't even try word lookups at all on phrases containing Cyrillic and just immediately fail out?

## Length limits, other possible constraints
* **Proposal:** capture the longest indexed word at index time and don't try to fuzzy-match words that are longer than `longest_word_length + edit_distance` at query time
* **Alternatives/considerations:**
    * The core consideration here is that our current algorithm for fuzzy-matching requires, at query time, generating all of the possible letter-dropped variants of a word in order to perform lookups, and that number grows with the length of the word, so there are potentially some pathological/DoS-y concerns if, say, a user submits a query with a 100-char-long word in it. Logically for words longer than the longest word in our index, it would take at least `query_word_lenght - longest_word_length` edits to successfully match, so we can fail if the word is longer than `longest_word_length + edit_distance`
    * One alternative would be to be even more aggressive, with a hard-coded or configurable threshold that might be shorter than the longest word (so, don't try to fuzzy-match 20-char-long words even if we have such words, and prefix-match them instead). Not sure this is worth the bother.

Another thing:
* **Maybe-proposal?:** don't try to fuzzy-correct one-letter words, which could translate to any other one-char-word (I'm also not even sure if this is supported in our fuzzy data structure, but I think it is?). We could also correct them, but only correct for the possibility that they're deletions (that is, should be two-letter words) and not substitutions, but our fuzzy-matching library doesn't currently allow for making this distinction.

**Comment by boblannon** [comment]
Numbers: totally makes sense, I think we don't lose much by missing `ma1n`
Length Limits: `longest_word_length + edit_distance` makes sense to me, and I'd prefer it to hard-coding a value. 

Alphabets: not clear on what the regex is doing. Are you saying that we'd only do the symspell process on things that match it?

**Comment by apendleton** [comment]
@boblannon right, we'd check if the word was comprised entirely of characters from our accepted alphabets, and attempt to fuzzy-match it if so, or otherwise only exact-match it (using the prefix graph instead of the fuzzy graph as it's faster for exact matches). We pretty much have to do some version of this (to avoid attempting to fuzzy-match Chinese), so I think the main questions are, how configurable vs. hard-coded is it, and which specific writing systems do we want to include or exclude.

**Comment by boblannon** [comment]
> Rather than hard-coding, we could make this configurable by the caller.

Meaning that carmen sends along a flag for `do_sympell`?  This is what I had imagined, since carmen is already implementing a ton of text-normalization and script detection. I imagine it will be less dev time to build on top of what we have there.  I think we've all agreed in other conversations that termops/tokenization/normalization logic is something we'd like to encapsulate, possibly in another library and possibly in rust. I think doing that would be easier if we kept those operations close together for the time being.

**Comment by apendleton** [comment]
If we go the configurable route, what I was thinking was that Carmen would supply a rule for what to include or not include at index build time, and we would store it and then apply it to incoming words at query time, so that they always match. Easiest is to hard-code. This is probably second-easiest. Carmen doesn't really do much in the way of script detection anymore so we'd need to bring that back.
